homeassistant/components/insteon_plm.py

```diff
@@ -6,115 +6,176 @@
 
     hass.data['insteon_plm'] = plm
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, plm.close)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, conn.close)
 
-    plm.protocol.devices.add_device_callback(async_plm_new_device, {})
+    plm.devices.add_device_callback(async_plm_new_device)
 
     return True
 
 
-def common_attributes(entity):
+def common_attributes(entity, state):
     """Return the device state attributes."""
-    attributes = {}
-    attributekeys = {
-        'address': 'INSTEON Address',
-        'description': 'Description',
-        'model': 'Model',
-        'cat': 'Category',
-        'subcat': 'Subcategory',
-        'firmware': 'Firmware',
-        'product_key': 'Product Key'
+    attributes = {
+        'INSTEON Address': entity.address.human,
+        'Description': entity.description,
+        'Model': entity.model,
+        'Category': '{:02x}'.format(entity.cat),
+        'Subcategory': '{:02x}'.format(entity.subcat),
+        'Product Key / Firmware': '{:02x}'.format(entity.product_key),
+        'Group': '{:02x}'.format(state.group)
     }
+    return attributes
 
-    hexkeys = ['cat', 'subcat', 'firmware']
 
-    for key in attributekeys:
-        name = attributekeys[key]
-        val = entity.get_attr(key)
-        if val is not None:
-            if key in hexkeys:
-                attributes[name] = hex(int(val))
-            else:
-                attributes[name] = val
-    return attributes
+State = collections.namedtuple('Product', 'stateType platform')
+
+
+class IPDB(object):
+    """Embodies the INSTEON Product Database static data
+and access methods."""
+
+    from insteonplm.states.onOff import (OnOffSwitch,
+                                         OnOffSwitch_OutletTop,
+                                         OnOffSwitch_OutletBottom,
+                                         OpenClosedRelay)
+
+    from insteonplm.states.dimmable import (DimmableSwitch,
+                                            DimmableSwitch_Fan)
+
+    from insteonplm.states.sensor import (VariableSensor,
+                                          OnOffSensor,
+                                          SmokeCO2Sensor,
+                                          IoLincSensor)
+
+    states = [
+        State(OnOffSwitch_OutletTop, 'switch'),
+        State(OnOffSwitch_OutletBottom, 'switch'),
+        State(OpenClosedRelay, 'switch'),
+        State(OnOffSwitch, 'switch'),
+
+        State(IoLincSensor, 'binary_sensor'),
+        State(SmokeCO2Sensor, 'sensor'),
+        State(OnOffSensor, 'binary_sensor'),
+        State(VariableSensor, 'sensor'),
+
+        State(DimmableSwitch_Fan, 'fan'),
+        State(DimmableSwitch, 'light')
+        ]
+
+    def __init__(self):
+        self.log = logging.getLogger(__name__)
+
+    def __len__(self):
+        return len(self.states)
+
+    def __iter__(self):
+        for product in self.states:
+            yield product
+
+    def __getitem__(self, key):
```

Comment:
Why not make it a dictionary if you want to support `getitem`?

Reply:
My concern about making this a dictionary is the key is a class which may not be hashable. Given that the list is small (and it will never get much bigger than 20) I don't believe there is a performance issue here. Unless you have strong objection, I prefer to leave this as is.